pkg/validator/shoot_validator.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	gcpvalidation "github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/validation"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/gardener/gardener/pkg/apis/core"
+)
+
+// getAllowedRegionZonesFromCloudProfile fetches the set of allowed zones from the Cloud Profile.
+func getAllowedRegionZonesFromCloudProfile(shoot *core.Shoot, cloudProfile *gardencorev1beta1.CloudProfile) sets.String {
+	shootRegion := shoot.Spec.Region
+	for _, region := range cloudProfile.Spec.Regions {
+		if region.Name == shootRegion {
+			gcpZones := sets.NewString()
+			for _, gcpZone := range region.Zones {
+				gcpZones.Insert(gcpZone.Name)
+			}
+			return gcpZones
+		}
+	}
+	return nil
+}
+
+func (v *Shoot) validateShoot(ctx context.Context, shoot *core.Shoot) error {
+	var (
+		fldPath            = field.NewPath("spec", "provider")
+		networkPath        = field.NewPath("spec", "networking")
+		infraConfigFldPath = fldPath.Child("infrastructureConfig")
+	)
+
+	// Network validation
+	if errList := gcpvalidation.ValidateNetworking(shoot.Spec.Networking, networkPath); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	// Provider validation
+	if shoot.Spec.Provider.InfrastructureConfig == nil {
+		return field.Required(infraConfigFldPath, "InfrastructureConfig must be set for GCP shoots")
+	}
+
+	infraConfig, err := decodeInfrastructureConfig(v.decoder, shoot.Spec.Provider.InfrastructureConfig, infraConfigFldPath)
```

Comment:
If you decode resources like `InfrastructureConfig` and `ControlPlaneConfig` here then they will be decoded multiple times in case of an update. I once had this as well for the other providers but revised it with a `validationContext` object which is passed to the `validateShoot ` function. https://github.com/gardener/gardener-extension-provider-openstack/blob/2f3220c4b77214e4c4db3279fac0a5a8c87dff2e/pkg/validator/shoot_validator.go#L53